### PR TITLE
py/objstr: Skip whitespace in bytes.fromhex().

### PR DESCRIPTION
**Pull request by glenn20**
## Skip whitespace characters between pairs of hex numbers in `bytes.fromhex()`

This makes `bytes.fromhex()` compatible with cpython. 

If whitespace char is encountered between pairs of hex numbers (specifying an octet), they will be skipped.
If a whitespace char is encountered **within** a pair of hex numbers specifying an octet, a ValueError will be raised.

Eg:

```py
>>> bytes.fromhex("abcdef")
b'\xab\xcd\xef'
>>> bytes.fromhex("ab cd ef")
b'\xab\xcd\xef'
>>> bytes.fromhex("ab c d ef")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: non-hex or odd number of hex digits found
>>> bytes.fromhex("ab cd\t\n    ef ")
b'\xab\xcd\xef`
```

Also extends `tests/basic/builtin_str_hex.py` to include tests with whitespace.

### CODE REVIEW

**Comment by glenn20** [comment]
After a few little tweaks, the code size is a little smaller and seems to be working fine (including other extended tests).

**Comment by dpgeorge** [comment]
I really like the way you've tried to make this a +0 code size change, in particular by combining the error message.

But unfortunately it lets in invalid hex chars, eg:
```py
>>> bytes.fromhex(':; <= >? @W XY Z[ \\] ^_ ``')
b'4Vx\x90\x124Vx\x99'
```
This is because of the arithmetic in `unichar_xdigit_value()`.

**Comment by dpgeorge** [comment]
@glenn20 can you think of a way to prevent invalid hex chars (see my comment above)?

**Comment by glenn20** [comment]
Whoops. Now, that one fell through the cracks. I'll take a look.

**Comment by glenn20** [comment]
Rebased and check for non-hex digits. Increases code size by 48 bytes on esp32.

Added tests for invalid chars and odd number of digits in tests/basics/builtin_str_hex.py.

File: py/objstr.c [hunk 1]
```diff
@@ -2014,27 +2014,21 @@ mp_obj_t mp_obj_bytes_fromhex(mp_obj_t type_in, mp_obj_t data) {
...
+        if (unichar_isspace(ch1 = *in++)) {
+            continue;  // Skip whitespace between hex digit pairs
         }
-        if (i & 1) {
-            hex_byte <<= 4;
-        } else {
-            *out++ = hex_byte;
-            hex_byte = 0;
+        if (in == in_end || !unichar_isxdigit(ch1) || !unichar_isxdigit(ch2 = *in++)) {
+            mp_raise_ValueError(MP_ERROR_TEXT("non-hex or odd number of digits"));
```
Comment (dpgeorge) [inline py/objstr.c:2027]: I think we can make this error message shorter, just use "non-hex digit".

CPython itself doesn't have a separate message for odd-length strings, it just has a single error:
```py
Python 3.12.3 (main, Apr 23 2024, 09:16:07) [GCC 13.2.1 20240417] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> bytes.fromhex('abcd')
b'\xab\xcd'
>>> bytes.fromhex('abc')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: non-hexadecimal number found in fromhex() arg at position 3
>>> bytes.fromhex('abc ')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: non-hexadecimal number found in fromhex() arg at position 3
>>> bytes.fromhex('ab c ')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: non-hexadecimal number found in fromhex() arg at position 4
>>> bytes.fromhex('ab cd')
b'\xab\xcd'
```

Comment (glenn20) [reply py/objstr.c:2027]: Ah yes. Done.

Comment (glenn20) [reply py/objstr.c]: And pushed fix to tests as well.

**Comment by dpgeorge** [comment]
Thanks for updating.  This is now a really nice change, improving CPython compatibility and effectively a zero code size increase.
